SalesforceCommerceCloud/pwa-kit
# Queries on hydration

For more of feature-parity with `getProps`, this PR would make it the default behaviour that queries do not get run during hydration. Because they already get run on the server-side.

[Ticket](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE0000172lUtYAI/view)

TLDR: The PR essentially implements the changes by:
- configuring react-query
- and managing the configuration options (such that the scope of the change is isolated to the server-side and hydration)

# Configuring react-query

Initially, I was exploring with increasing the `staleTime` value (because by default it's `0`). Adam had [suggested](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/734#discussion_r982589443) it last time. And it was also mentioned in the react-query [documentation](https://tanstack.com/query/v4/docs/guides/ssr#staleness-is-measured-from-when-the-query-was-fetched-on-the-server). However, while it does work well, I wasn't sure what the appropriate value for it is. 

But it turned out that there's another option that also does what we want. It's the `refetchOnMount` boolean. On hydration, the page component mounts, and there will be a query refetch in the background. With `refetchOnMount: false`, there will be no more refetch during hydration.

## Refetch vs Retry

Aside: it's important to differentiate between a refetch and a retry in the react-query world. While both repeat the existing queries:
- **refetch** is for _successful_ queries
  - Refetch happens in the _background_. It's useful to think of it in terms of [stale-while-revalidate](https://www.reddit.com/r/reactjs/comments/ncfohm/comment/gy8odwn/?utm_source=reddit&utm_medium=web2x&context=3) approach. After a successful query, it will be automatically categorized as _stale_, and thus React Query would like to keep it up-to-date (see this [article](https://tkdodo.eu/blog/react-query-as-a-state-manager#smart-refetches) for the times when it refetches). Over time, any access to this query would result in: first, fetching the stale data from cache, and then refetch the latest data in the background.
- **retry** is for _failed_ queries
  - Retry is _not_ done in the background. There is no available data yet, and so it's like another attempt at running the query.

# How to manage the configuration options

It's still to be decided how we want to manage the configuration options: those options that the users would pass in and also the ones that the pwa-kit SDK relies on. In this PR, I propose that we draw the line between server and client side:
- our SDK options applied during the server side and the hydration on client-side
- after hydration, the user configuration would be applied

I think the above setup is reasonable, given that react-query is optimized and meant for client-side. 

I have some more thoughts on managing the configuration options in this [Quip doc](https://salesforce.quip.com/aI6BAvLQH60i).

# How to Test-Drive This PR

You can go to the test-commerce project and verify what happens now during hydration:

https://user-images.githubusercontent.com/847300/196304108-c0f321de-80c3-4104-8192-d04b656e6ef0.mp4
